timemachine/md/exchange/exchange_mover.py

```diff
@@ -355,8 +355,29 @@ def swap_vi_into_vj(
         log_weights_after_full = np.array(log_weights_after_full)
         log_weights_after = log_weights_after_full[vj_plus_one_idxs]
 
+        # directionality is going from region i into region j
+        # take care of asymmetric boundary conditions
+        if len(vi_mols) == self.num_waters:
```

Comment:
Sorry, not immediately actionable, but in a later pass I'd like to figure out how to write this in terms of a more direct comparison `len(vi_mols) == 0` or similar.

---------

Want the code to explicitly reflect that we first make a discrete random choice of whether we're doing a "delete from i" move or a "delete from j" move, where the distribution over choices is conditioned on x.

f_i(x) = is_there_at_least_one_water_in_i(x)
f_j(x) = is_there_at_least_one_water_in_j(x)

f(x) = (f_i(x), f_j(x))
move_probs(x) = f(x)  / sum(f(x))

i.e., we sample move_idx ~ Categorical(move_probs(x0)),

then we produce x_prop ~ moves[move_idx](. | x0)

... following the recipe in https://www.pnas.org/doi/full/10.1073/pnas.1106094108 (where the random choice of "protocol" can depend on x).

The current implementation is a special case where the distribution over choices is uniform over all currently feasible choices. (i.e. uniform over whichever regions contain at least one water)

Should be equivalent to the nested if statement definition, but might reduce the number of special cases that need to be listed, and might be more easily extended to the case of multiple regions, non-uniform move selection probabilities, or other directions...

Reply:
nice! I like that formulation and it makes sense. Will mark this to be cleaned-up when we move away from the special case (eg. more regions, non-uniform choices, et.c).

Reply:
> Sorry, not immediately actionable, but in a later pass I'd like to figure out how to write this in terms of a more direct comparison len(vi_mols) == 0 or similar.

Made a change in https://github.com/proteneer/timemachine/pull/1228/commits/d23078bd896db9f02a8100e4424f6d397529ef05

That should reflect this better. Also helped identify another corner case when `N_waters == 1` (which symmetrizes the probabilities again). The original intent was to deal with asymmetries in `N_waters > 1`  but `N_waters_in_vol_i`, `N_waters_in_vol_j` could be `0` or `1`